### PR TITLE
[15 min fix] Enforce an upper limit for articles.body_markdown

### PR DESCRIPTION
**Pull request by rhymes**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Adds an upper limit to `Article.body_markdown`

## Related Tickets & Documents

Closes https://github.com/forem/internalEngineering/issues/410

## QA Instructions, Screenshots, Recordings

1. load the console and load an article
2. `article.body_markdown = 'a' * 801.kilobytes`
3. `article.validate!`
4. get the error `ActiveRecord::RecordInvalid: Validation failed: Body markdown is too long (maximum is 819200 characters)`
5. `article.body_markdown = 'a' * 800.kilobytes`
6. `article.validate!` should yield `true`

## Added tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [x] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

### CODE REVIEW

**Review by mstruve — COMMENTED** [review]
Are we OK leaving a bunch of unpublished articles invalid? They won't be able to be saved then until they are updated which I could see possibly leading to some problems for future DUS's where we would assume we could save everything.

**Review by vaidehijoshi — APPROVED** [review]
Should we write a changelog post for this? 🤔 Maybe that's overkill, but I feel like at a minimum, maybe we can just update the Forem Admin Guide!

File: spec/models/article_spec.rb [hunk 1]
```diff
@@ -32,6 +32,7 @@ def build_and_validate_article(*args)
     it { is_expected.to have_many(:tags) }
     it { is_expected.to have_many(:user_subscriptions).dependent(:nullify) }
 
+    it { is_expected.to validate_length_of(:body_markdown).is_at_least(0).is_at_most(800.kilobytes) }
```
Comment (djuber) [inline spec/models/article_spec.rb:35]: how does this work with utf8 encoding? If I create an article 750,000 Chinese characters using multibyte entities, would length be in characters (unicode entities) or bytes? Is 3MB a safe length?

```
irb(main):005:0> ('𑨕' * 800 * 1024).length
=> 3276800
```
Just choosing Zanabazar square letter ttha as an example (0x11a15) here I get perhaps more than you anticipated.

If the issue with the database is _also_ expressed in characters and not bytes, feel free to dismiss this!

**Comment by vaidehijoshi** [comment]
> Are we OK leaving a bunch of unpublished articles invalid? 

That's a good question; how many unpublished articles do we have that exceed this cap? 🤔

**Comment by rhymes** [comment]
@mstruve 

> Are we OK leaving a bunch of unpublished articles invalid? They won't be able to be saved then until they are updated which I could see possibly leading to some problems for future DUS's where we would assume we could save everything.

I partialy considered it but I only thought about the user experience, by seeing the new error they'll be nudged to split the article in a series. It's true they'll be invalid if we resave them out of bounds...

@vaidehijoshi 

> Should we write a changelog post for this? 🤔 

good call, I'm going to add the tag

> That's a good question; how many unpublished articles do we have that exceed this cap? 🤔

11 draft aricles https://dev.to/admin/blazer/queries/417-articles-over-800-kb

**Comment by mstruve** [comment]
I think we need a way to make those invalid articles valid otherwise if someone goes to backfill a field that is needed by the UI, those articles can't be updated and suddenly we have 11 articles that will possibly break things. I am fine merging this if we have a plan to fix those articles, whether that's manually truncating them after reaching out to users or some other solution. Maybe we go with the valid after a date solution @vaidehijoshi used for mention counts.

File: spec/models/article_spec.rb (hunk 1 above)
Comment (rhymes) [reply spec/models/article_spec.rb:35]: Good point! The DB says `max 1048575 bytes`. From https://www.postgresql.org/docs/11/textsearch-limitations.html

> The length of a tsvector (lexemes + positions) must be less than 1 megabyte

Rails uses `string.length` to check https://github.com/rails/rails/blob/f95c0b7e96eb36bc3efc0c5beffbb9e84ea664e4/activemodel/lib/active_model/validations/length.rb#L41

**Comment by rhymes** [comment]
@mstruve I'd rather not add that `if`, what we did for the 2 published articles was to export the content, message the owners and then we took care of the articles. I think we can do something similar by contacting them with an export of the markdown and then we can trim the content. Best to double check the game plan with @michael-tharrington which helped @maestromac and I for those 2 articles that were published

**Review by mstruve — APPROVED** [review]
Went straight to @michael-tharrington and asked and he said he was good to help with that. Thanks for the game plan @rhymes!

**Comment by michael-tharrington** [comment]
Heyo!

We're totally happy to help out with contacting the 11 folks here (cc' @FounderCasey & @ellativity). Ideally we'd export the MD used in their posts and send it to them with advice to split their post up into multiple posts & use a series if they plan to publish.

Can anyone please help by providing us with these exported files? (cc' @rhymes as I realize you did this previously)

**Comment by rhymes** [comment]
@michael-tharrington I'm on it! :)

File: spec/models/article_spec.rb (hunk 1 above)
Comment (djuber) [reply spec/models/article_spec.rb:35]: is there a "validate bytesize of" function?

**Comment by michael-tharrington** [comment]
Dropping this in here so that @ellativity & @FounderCasey can more easily access it [csvs.zip](https://github.com/forem/forem/files/6320653/csvs.zip)

File: spec/models/article_spec.rb (hunk 1 above)
Comment (jgaskins) [reply spec/models/article_spec.rb:35]: @rhymes That'd be characters (Unicode codepoints). If we want bytes, we need to use `String#bytesize`.

Comment (jgaskins) [reply spec/models/article_spec.rb:35]: Also, why do we have specs on validations and associations like this? Is there any way to make this pass other than the exact line of code for the validation?

File: spec/services/articles/creator_spec.rb [hunk 2]
```diff
@@ -67,7 +67,7 @@
 
       expect(article.decorated?).to be(false)
       expect(article).not_to be_persisted
-      expect(article.errors.size).to eq(1)
+      expect(article.errors.size.positive?).to be(true)
```
Comment (jgaskins) [inline spec/services/articles/creator_spec.rb:70]: Why the change away from `eq(1)` here?

File: spec/models/article_spec.rb (hunk 1 above)
Comment (rhymes) [reply spec/models/article_spec.rb:35]: > That'd be characters (Unicode codepoints). If we want bytes, we need to use String#bytesize.

Ok we need a custom validator then. I think this is why Rails doesn't have a bytesize validator: https://github.com/rails/rails/pull/22870#issuecomment-169439090

We can use something like this https://github.com/rails/rails/issues/19570#issuecomment-348366536

> Also, why do we have specs on validations and associations like this? Is there any way to make this pass other than the exact line of code for the validation?

I'm not sure I understand this part 😊

Comment (rhymes) [reply spec/models/article_spec.rb:35]: For some reason I really thought Rails was using `.bytesize` internally 😭

File: app/models/article.rb [hunk 3]
```diff
@@ -54,7 +55,8 @@ class Article < ApplicationRecord
            inverse_of: :commentable,
            class_name: "Comment"
 
-  validates :body_markdown, length: { minimum: 0, allow_nil: false }, uniqueness: { scope: %i[user_id title] }
+  validates :body_markdown, length: { in: 0..MAX_BODY_MARKDOWN_SIZE, allow_nil: false }
+  validates :body_markdown, uniqueness: { scope: %i[user_id title] }
```
Comment (jgaskins) [inline app/models/article.rb:59]: A uniqueness validation here may not be a good idea. When Rails invokes a uniqueness validation, it'll run a query like this:

```sql
SELECT 1 AS one
FROM "articles"
WHERE "articles"."body_markdown" = $1
LIMIT $2
```

There is no index on `body_markdown`, so it has to check the column data. Because `body_markdown` can be pretty large, its content will almost always be held in [TOAST tables](https://wiki.postgresql.org/wiki/TOAST), adding another place to look on disk during the query, making this slower than a simple value lookup.

To be clear, I'm also not saying we should add an index to this. Indexing `body_markdown` would bloat the indexes — on DEV, this column alone is over 2GB and all other indexes combined on this table is currently 3GB.

If we want to ensure that double-clicking the submit button doesn't create duplicate articles, I'd recommend creating a `body_markdown_checksum` column, storing a hash of the `body_markdown` in it, and creating a unique index on that column.

**Review by jgaskins — CHANGES_REQUESTED** [review]
Blocking on the uniqueness validation on `body_markdown`. See comment below.

File: app/models/article.rb (hunk 3 above)
Comment (rhymes) [reply app/models/article.rb:59]: @jgaskins thanks for the feedback, though I think everything you said has been already covered. Let me know if we should change something

A bit of context:

the uniqueness validator already exists in production, I only moved it to a different line for readability:

https://github.com/forem/forem/blob/c842d1a0c6718f319f167e9ec13d4d7d5d6b6d39/app/models/article.rb#L57

We do have an index on `body_markdown` which uses a digest of the markdown:

https://github.com/forem/forem/blob/c842d1a0c6718f319f167e9ec13d4d7d5d6b6d39/db/schema.rb#L155

File: spec/services/articles/creator_spec.rb (hunk 2 above)
Comment (rhymes) [reply spec/services/articles/creator_spec.rb:70]: Because with `nil` body and `in:` you get both the min and the max errors

File: app/models/article.rb (hunk 3 above)
Comment (jgaskins) [reply app/models/article.rb:59]: Apologies, I misread the diff. I didn't notice this was an existing validation. This is what I saw for the changed line 😆 

Screenshot of the diff of the first line

The line wrapping had me all confused. 😞 

Since it's already live, I looked at some traces where this happens and it's surprisingly fast (1-2ms), apparently because it's not globally unique like I thought when I read it. Still, I don't _love_ the idea of checking the database for uniqueness on something that can be 800KB since it negates any benefit we were gaining from just checking the digest for uniqueness, but that's not a problem for this PR to solve.

Sorry, my brain's apparently still fried from RailsConf. 😄

File: spec/models/article_spec.rb (hunk 1 above)
Comment (jgaskins) [reply spec/models/article_spec.rb:35]: > > Also, why do we have specs on validations and associations like this? Is there any way to make this pass other than the exact line of code for the validation?
> 
> I'm not sure I understand this part 😊

As far as I know, the only way to make `expect(Article).to validate_length_of(:body_markdown)...` pass is to have `validates :body_markdown, length: { ... }` in `Article`. It specifies the implementation of the validation rather than the interface.

We could test the interface by saying something like this:

```ruby
article = build(:article)

article.body_markdown = '.' * 800.kilobytes
expect(article).to be_valid

article.body_markdown << '.'
expect(article).to be_invalid
expect(article.errors[:body_markdown]).to include "is too long"

article.body_markdown = '❤️' * (800.kilobytes / '❤️'.bytesize) # 819_198 out of 819_200 allowed bytes
expect(article).to be_valid

article.body_markdown << '❤️' # 819_204 bytes, exceeds 819_200 allowed
expect(article).not_to be_valid
expect(article.errors[:body_markdown]).to include "is too long"
```

This way, it doesn't matter how we change the implementation of the validation as long as it conforms to the invariant.

**Comment by rhymes** [comment]
@jgaskins I changed the code to use a custom `bytesize` validator :)

File: app/models/article.rb [hunk 4]
```diff
@@ -54,7 +54,9 @@ class Article < ApplicationRecord
            inverse_of: :commentable,
            class_name: "Comment"
 
-  validates :body_markdown, length: { minimum: 0, allow_nil: false }, uniqueness: { scope: %i[user_id title] }
+  validates :body_markdown, bytesize: { maximum: 800.kilobytes, too_long: "is too long." }
```
Comment (rhymes) [inline app/models/article.rb:57]: the custom message is because by default Rails would use `too many characters` but it would confuse people as bytesize is not the same as characters, and nobody actually checks how many characters they have typed :D We could improve on the messaging but I think given its an edge case, we could live with this for the moment

File: spec/services/articles/creator_spec.rb (hunk 2 above)
Comment (jgaskins) [reply spec/services/articles/creator_spec.rb]: Right, so if only one of those validations fails here, this test still passes. Was that intentional?

Comment (rhymes) [reply spec/services/articles/creator_spec.rb]: I reverted it now that I've moved the check to `bytesize` and a separate validator

File: spec/models/article_spec.rb (hunk 1 above)
Comment (rhymes) [reply spec/models/article_spec.rb]: Ah, got it! 

I think we're covered. Right now we have a length validator:

https://github.com/forem/forem/blob/9fe7be190bd364f48deb8895f845378a2c9bfa27/app/models/article.rb#L57

which is now tested by 

```ruby
it { is_expected.to validate_length_of(:body_markdown).is_at_least(0) }
```

and after the latest change we'll also have a `bytesize` validator and its test

File: spec/models/article_spec.rb [hunk 5]
```diff
@@ -81,11 +82,25 @@ def build_and_validate_article(*args)
...
+        article.body_markdown = "𑨕" * 204_800 # 4 bytes x 204800 = 800 kilobytes
+
+        expect(article).to be_valid
+      end
+
+      it "is not valid if its bytesize exceeds 800 kilobytes" do
+        article.body_markdown = "𑨕" * 204_801
+
+        expect(article).not_to be_valid
+        expect(article.errors_as_sentence).to match("too long")
```
Comment (jgaskins) [inline spec/models/article_spec.rb:103]: TIL `errors_as_sentence`!

File: app/validators/bytesize_validator.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
+# Adapted from https://github.com/rails/rails/issues/19570#issuecomment-348366536
+# and https://github.com/rails/rails/blob/v6.1.3.1/activemodel/lib/active_model/validations/length.rb
+
+# Currently only supports `:maximum`
+class BytesizeValidator < ActiveModel::EachValidator
+  MESSAGES = { maximum: :too_long }.freeze
+  CHECKS = { maximum: :<= }.freeze
```
Comment (jgaskins) [inline app/validators/bytesize_validator.rb:7]: This is a great move to make it easy to extend!

**Review by jgaskins — APPROVED** [review]
This ended up super clean!
